Feature/add utils functions

guilhermebruzzi opened the pull request:
This is not a breaking change PR (meaning that the same import of parcelify and the same tests on master pass).

But it adds another criteria for parcelify named groupByAvailableDeliveryWindows (default: false) that will allow to construct a better interface for selecting scheduled delivery items.

This PR also break the existing functions in multiple files, add individual tests to them all and add more functions the only the omnishipping module was using (little drawer).

Review with love and sorry for the size of this PR! <3

guilhermebruzzi committed:
Refactor of functions on more friendly named modules

guilhermebruzzi committed:
Add more functions and tests

guilhermebruzzi committed:
Improve testing mocks

guilhermebruzzi committed:
Init better sla tests and utils

guilhermebruzzi committed:
Add getSelectedSla test

guilhermebruzzi committed:
Add findSlaWithChannel test

guilhermebruzzi committed:
Add getSelectedSlaInSlas tests

guilhermebruzzi committed:
Add hasDeliveryWindows tests

guilhermebruzzi committed:
Add hasSLAs tests

guilhermebruzzi committed:
Add delivery channel tests

guilhermebruzzi committed:
Initial scheduled delivery refactor and better null treating on logisticsInfo and Address

WIP: scheduled-delivery utils

Null treating on logisticsInfo and Address

WIP: WIP: scheduled-delivery utils 2

guilhermebruzzi committed:
Add Shipping and items initial tests

guilhermebruzzi committed:
Add more items tests

guilhermebruzzi committed:
Add getSelectedSlaIfMatchSlaId on sla package and add tests to each module related

guilhermebruzzi committed:
Add more scheduled delivery tests

Add selectDeliveryWindow and tests to it

Add getScheduledDeliverySLA test

Add getFirstScheduledDelivery test

guilhermebruzzi committed:
Add more shipping tests

guilhermebruzzi committed:
Add shipping extra tests

guilhermebruzzi committed:
Replace filter for find  or some and add polyfill to find

guilhermebruzzi committed:
Add more shipping tests

guilhermebruzzi committed:
Fix getNewLogisticsInfoWithSelectedScheduled and add tests to it

guilhermebruzzi committed:
Add group by availableDeliveryWindow criteria and tests

guilhermebruzzi committed:
Refactor on selectDeliveryWindow to use selectedSla name and checkIfHasDeliveryWindow to use areDeliveryWindowsEquals

guilhermebruzzi committed:
Refactor on getScheduledDeliverySLA param name

guilhermebruzzi committed:
Add on README referrence to the new functions organization

klzns commented on src/polyfills.js:
```diff
@@ -0,0 +1,29 @@
+if (!Array.prototype.find) {
```
Acho que nao é papel da lib colocar o Polyfill. 

Acho que so instruir o client de colocar eles mesmo.

No OMS eu coloquei:

```html
<script src="https://cdn.polyfill.io/v2/polyfill.min.js?features=Array.prototype.find|gated"></script>
```

klzns replied:
Talvez colocar as instruções no README.

guilhermebruzzi replied:
@klzns Como são muito pequenos, você não acha que vale a pena? Porque só com eles já é garantido IE9+ em todas as funções sem quem tiver que chamar não precisar incluir mais nada (é 427B esse arquivo minificado). To pensando na developer XP que quem usa! hehehe

klzns replied:
Entendo, mas no render, por exemplo, esses polyfills ja sao inseridos.

guilhermebruzzi replied:
Eu penso que quando tudo usar o render, esse tipo de detalhe realmente estará resolvido, mas ao meu ver por enquanto vale deixar ali mesmo. Se considerar GZIP estamos falando de 283B e evitar um request a toa

omninando replied:
Eu concordo que vale a pena incluir nem que precisarmos retirar isso no futuro. É uma adição sem custo alto e que dá um ganho em suporte bem grande

analuizamtg replied:
concordo que deve ser incluído sim. infelizmente ainda temos poucas coisas no render!

klzns commented on src/index.js:
```diff
@@ -1,348 +1,3 @@
-    ? logisticInfo.slas.find(sla => sla.id === selectedSla)
-    : null
-}
+export default parcelify
```
Acho que a função parcelify podia ta aqui.

guilhermebruzzi replied:
@klzns Eu acho que é uma boa falarmos sobre o parcelify mesmo, pra mim é um nome peculiar pra uma função que separa itens em pacotes de entrega hehehe (eu provavelmente não estou com o contexto certo do porque do nome e como ela foi criada na real).
Foi por isso que coloquei em package.js

klzns replied:
@guilhermebruzzi antes era `package` mas no OMS a API ja tem uma propriedade do pedido chamada `package` o que causava confusão pq não dava pra saber quando estavamos nos referindo ao `package` da API ou o `package` da lib. 

Preferimos mudar o nosso termo para `parcel` e evitar essa confusão. O termo `package` é apenas usado para se referir a entidade da API.

guilhermebruzzi replied:
Só pra deixar aqui que no fim ficou em parcel.js

klzns commented on src/package.js:
```diff
@@ -0,0 +1,182 @@
+import { getShippingEstimateQuantityInSeconds } from '@vtex/estimate-calculator'
```
O nome desse arquivo `package.js` é meio confuso.

guilhermebruzzi replied:
@klzns É parecido com o meu comentário acima, do porque eu não deixei nada em index.js que não fosse o export default (que foi mais pra garantir a retrocompatibilidade)

klzns commented on README.md:
```diff
@@ -260,6 +264,96 @@ parcelify(order, { criteria: { seller: false } })
+
+All functions also import some [polyfills](./src/polyfills) worth checking.
+
+### getDeliveryChannel
```
Mostrar como importar essas funções.

Em que path elas estão disponiveis?

Exemplo:

```js
import { getDeliveryChannel } from '@vtex/delivery-packages/dist/shipping'
```

Nem sem como o rollup vai gerar esses arquivos. É bom ver se ta td certo com o build.

guilhermebruzzi replied:
@klzns Vou conferir isso do build e incluir nos exemplos o import! :)

klzns reviewed (changes requested):
**Todas** as funções agora estão sendo exportadas. Elas agora fazem parte da API pública?

Isso torna a API gigante. Tem algum valor realmente deixar tudo isso público?

Podia ter alguma separação na estrutura de arquivos pra ficar claro o que é publico e o que é privado. Alguem pode fazer um refactoring nesse pacote e acabar quebrando a API sem querer, por não saber que é um metodo publico.

Tambem se certificar que todas as funções que são API publicas tenham testes.

guilhermebruzzi commented:
@klzns Todas as funções com `export` e que estão no REDME tem testes. Vou garantir que essa separação esteja mais clara nos arquivos e que todas as funções auxiliares que talvez não tenham teste, não tenham export tbm

guilhermebruzzi committed:
Separating private and public functions

guilhermebruzzi committed:
Fix hasDeliveryWindows typo on shipping

guilhermebruzzi committed:
Fix build for multiple export files format

guilhermebruzzi committed:
Detailing on README how to import all public auxiliary functions

guilhermebruzzi commented:
@klzns O build com rollup tá exportando certo agora, todas as funções publicas estão explicitadas nos arquivos e com testes, no README tem exemplo de como importá-las com a mesma sintaxe que tinha do parcelify e o parcelify e funções auxiliares dele eu coloquei então num arquivo chamado `parcel.js`, com o `index.js` existindo para retrocompatibilidade.
A ideia é tudo relacionado a delivery, a gente ir adicionando aqui depois desse PR mergeado e parar de criar em cada projeto utils para resolver esses mesmos problemas

klzns commented on README.md:
```diff
@@ -260,6 +264,124 @@ parcelify(order, { criteria: { seller: false } })
 
+## Other important functions
+
+See [src](./src/) folder to see all available functions and [tests](./tests/) folder for more examples.
```
Eu tiraria isso. A API tem que ta listada no README msm.

klzns commented on README.md:
```diff
@@ -260,6 +264,124 @@ parcelify(order, { criteria: { seller: false } })
+
+See [src](./src/) folder to see all available functions and [tests](./tests/) folder for more examples.
+
+All functions also import some [polyfills](./src/polyfills) worth checking.
```
Tiraria, detalhe de implementacao

klzns commented on README.md:
```diff
@@ -260,6 +264,124 @@ parcelify(order, { criteria: { seller: false } })
+
+### getDeliveryChannel
+
+**import:** const { getDeliveryChannel } = require('@vtex/delivery-packages/delivery-channel')
```
Eu criaria sessões para cada módulo e melhoraria a descrição dos parâmetros. Acho que em uma lib de funções isso é o mais importante, se não fica difícil saber se a função faz o que se quer.

Desculpa ta empacando o PR, mas eu acho que isso é real importante 🙏 

Ex:


## Delivery Channel 
> @vtex/delivery-packages/delivery-channel

### getDeliveryChannel(deliveryChannelSource)

Get the delivery channel blablabla. Small description.

##### Usage
```js
const { getDeliveryChannel } = require('@vtex/delivery-packages/delivery-channel')

getDeliveryChannel({ deliveryChannel: 'pickup-in-point'})
// -> 'pickup-in-point'
```

**params:**
- **deliveryChannelSource**
Type: `object` or `string`
Small description of param

**returns:**
- **deliveryChannel**
Type: `string`
Small description of param

### isPickup(deliveryChannelSource)

Get the delivery channel blablabla. Small description.

##### Usage
```js
const { isPickup } = require('@vtex/delivery-packages/delivery-channel')

isPickup({ deliveryChannel: 'pickup-in-point'})
// -> true
```

**params:**
- **deliveryChannelSource**
Type: `object` or `string`
Small description of param

**returns:**
- **Boolean**
Type: `boolean`
If it is or it's not pickup point.


## Items
> @vtex/delivery-packages/items

### getNewItems(items, changes)

Get the new items blablabla. Small description.

##### Usage
```js
const { getNewItems } = require('@vtex/delivery-packages/delivery-channel')

getNewItems(items, changes)
// -> 'pickup-in-point'
```

**params:**
- **items**
Type: `array` of `object`
Array of items
- **changes**
Type: `array` of `object`
Array of items

**returns:**
- **deliveryChannel**
Type: `array`
The items received with the changes (itemsAdded and itemsRemoved) passed

klzns replied:
O trabalho de documentar também é bom pra refletir a API.

guilhermebruzzi committed:
Update package-lock

guilhermebruzzi committed:
Add getSlaObj as extra function

klzns commented on README.md:
```diff
@@ -260,6 +264,124 @@ parcelify(order, { criteria: { seller: false } })
+
+### getNewItems
+
+**import:** const { findChannelById } = require('@vtex/delivery-packages/items')
```
Opa, nome da funcao ta errado

guilhermebruzzi committed:
Fix README and prepare for new format

guilhermebruzzi committed:
Add Better helper functions documentation and improve delivery channel functions params

Enhance findChannelById deliveryChannel param to accept any deliveryChannel source

WIP: Better README

Add better functions examples and documentation on README

omninando commented on README.md:
````diff
@@ -260,6 +264,1009 @@ parcelify(order, { criteria: { seller: false } })
+```js
+const { isPickup } = require('@vtex/delivery-packages/delivery-channel')
+
+isPickup({ id: 'pickup-in-point'})
````
eu adicionaria os outros usos do isPickup tbm tipo 

```
isPickup({ selectedDeliveryChannel: 'pickup-in-point' })
// true

isPickup({ deliveryChannel:  'pickup-in-point' })
// true
```

omninando commented on README.md:
````diff
@@ -260,6 +264,1009 @@ parcelify(order, { criteria: { seller: false } })
+```js
+const { isDelivery } = require('@vtex/delivery-packages/delivery-channel')
+
+isDelivery({ id: 'pickup-in-point'})
````
acho que a mesma coisa aqui

```
isDelivery({ selectedDeliveryChannel: 'delivery' })
// true

isDelivery({ deliveryChannel:  'delivery' })
// true
```

guilhermebruzzi committed:
Add more examples for isDelivery and isPickup

guilhermebruzzi committed:
Add postreleasy command

guilhermebruzzi committed:
Release v2.3.0
